### ISSUE TITLE
JSON summary field types

### ISSUE THREAD

**Issue by marian-babik** [perfsonar/toolkit#248]
While trying to insert json summaries for the WLCG community perfSONARs into ElasticSearch we have found two issues that we think would be of general interest:
1. Some of the fields are loosely typed, for example network speed field is an integer, but can change to string in case it's unknown. Others are strings while it would probably make sense to have them as integers (like number of cores and cpus). This causes a lot of issues in parsing and imports to ElasticSearch, which expects field types to be consistent. It would be great if you could review all fields and make them consistent (if something is int then ensure it stays int) as this would greatly simplify parsing and imports in general. Btw. specifically for ES, it's best to simply remove fields that are unknown.

2. This one is related to the way ElasticSearch works. In general, ES expects documents with static keys, so for the case of JSON summaries, it's fine to add/remove keys, but key names should be fixed across documents. For example, in 4.0 "interfaces" were introduces like this:
```
"interfaces": [
{
    "hostnames": {
          "10.10.128.35": [
                "psmsu05.msulocal"
]
},
```
Using IP address as a key implies that you have a dynamic key name, which causes an explosion in ES indexing (which usually brings down ES). It would be better to re-structure it to keep keys static across the document, e.g.:
```
"interfaces": [
{
    "hostnames": [
     {
          "hostname" : "psmsu05.msulocal",
          "ip_address" : "10.10.128.35"
     }, ... ]
   }, 
or simply just:
    "hostnames": ["psmsu05.msulocal", ... ]
```

**Comment by arlake228** [comment]
In the meantime, it might be worth looking at JQ to transform it before you parse it.

**Comment by marian-babik** [comment]
Thanks, btw. collector we use to insert to ES is at:
https://github.com/ATLAS-Analytics/NetworkWeatherService/blob/master/NetworkWeatherCollector/psMetaCollector.py
(m holds the summary json)

We don't really want to parse/change the toolkit json summary, just want to insert it as it is, letting ES deal with the structure, which is what it does quite well. 

So what is done now to remedy existing issues is that we remove the entire section with interfaces and then run a recursive clean up function to remove all fields with none or 'unknown' values (we're still in the process of discovering what other values we get that won't work).

**Comment by marian-babik** [comment]
Clean up code is not yet merged - still in the pull request (https://github.com/ATLAS-Analytics/NetworkWeatherService/pull/10)

**Comment by apertome** [comment]
I think this is not going to happen for 4.1, but if @marian-babik could provide an example of what they want the data to look like, that would help me more than trying to read the python code. I think I can envision it, but I want to make sure we're on the same page.

I'm also not sure what/who relies on this data for something, and I don't want to break existing integrations, so I want to do this carefully. What if we added a "format" parameter and you could append `&format=staticfieldnames` or `&version=2` or something like that in the URL. How inconvenient would that be?

Then we could suggest people who use the JSON format to try using that instead, and once we fix any issues, we make the new format the default.

In some cases being able to know explicitly that a value is unknown or was not found is useful information, rather than it simply not being in output data, but your comments about data types not being consistent are well taken.

**Comment by marian-babik** [comment]
Creating a version or adding any other parameters to the URL to get alternative output is no problem.

Concerning unknowns, I see two options, one is to impose the field to be string and convert int/floats to string, the other is to have 0 or 0.0 (or -1) for unknown and keep the attribute in int/float. None of them are perfect, but I think mixing the data types for a given attribute must be causing more issues than it solves as anyone parsing this needs to have a special branch to catch the this.

**Comment by mfeit-internet2** [comment]
`null` would be a better choice, as would making the pair optional and omitting it if there's no value.  Using special values that are otherwise valid causes all sorts of problems.

**Comment by apertome** [comment]
@marian-babik would it be possible to provide an example of the JSON format after you've done your conversion? Thanks!

**Comment by apertome** [comment]
After discussing this further with Marian, the example he provided above is really the only part that needs to change, in terms of the dynamic keys.

Here is an example, but all toolkits have this JSON output available:
http://perfsonar-dev9.grnoc.iu.edu/toolkit/services/host.cgi?method=get_summary

To implement this, we need to do a couple things:

- [ ] Add an optional parameter to host.cgi called `version`
- [ ] If `version` is 1, the output should match the current output without any `version` specified
- [ ] When called with `&version=2`, the output should be modified in these ways
  - [ ] 1.`"unknown"` values (this is a string) should be `null` instead
  - [ ] 2. All numeric values should be output as numbers, not strings. i.e. `1500` rather than `"1500"`
  - [ ] 3. Boolean fields should be `true` or `false` rather than `"yes"` or `"1"` ( this applies to `enabled`, `is_running`, and `is_vm`, at the very least)
  - [ ] 4. The `interfaces` object needs to be changed so it does not use dynamic keys. For example:

```
{
"hostnames": {
"140.182.49.154": [
"perfsonar-dev9.grnoc.iu.edu"
]},
"ipv4_address": [
"140.182.49.154"
],
"iface": "eth0",
"ipv6_address": [],
"mac": "fa:16:3e:70:0d:8c",
"mtu": 9000
},
```

Should be simplified to this (notice all the fields are arrays, in case there are multiple addresses)
```
"addresses": {
     "ipv4_address": [ "140.182.49.154" ],
    "hostname": [ "perfsonar-dev9.grnoc.iu.edu" ],
    "ipv6_address": []

}
```

Once this is done, it would be good to let @marian-babik and/or @ShawnMcKee review and make sure what we have meets their needs. We'll need a separate ticket for adapting the toolkit to use the new format, although for now let's keep version 1 the default
